### DISCUSSION TITLE
Get "The specified key does not exist" error while trying to Import data with the /v2/vectordb/jobs/import/create API

### DISCUSSION THREAD

**Discussion by liorf95** [milvus-io/milvus#36653] [Q&A and General discussion]
Trying 
curl --request POST "http://xxxxx:19530/v2/vectordb/jobs/import/create" \
--header "Content-Type: application/json" \
--data-raw '{
    "files": [
        [
            "/31ad591e-97c9-42c6-9c68-94daa43eea8a/1.parquet"
        ]
    ],
    "collectionName": "quick_setup"
}'
 Always return "The specified key does not exist." for any parquet or json files.

What key do I need?

**Comment by xiaofan-luan** [comment]
where did you put the file?
milvus is assuming the file is on minio

**Reply by liorf95** [reply]
I put the file exactly in the same place where the RemoteBulkWriter does - on the minio bucket mounted root path.
The difference I can see is that the RemoteBulkWriter uses to create a 1.parquet **folder** contains several hundreds part files and a xl.meta file and not just a 1.parquet file as the LocalBulkWriter creates.

BTW - I see that there is already an open bug regard this issue:
https://github.com/milvus-io/milvus/issues/35530

**Comment by yhmo** [comment]
@liorf95 
Let's say we have a collection with this schema:
```
fields = [
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True),
    FieldSchema(name="vector", dtype=DataType.FLOAT_VECTOR, dim=4),
]
schema = CollectionSchema(fields=fields)
```

By using RemoteBulkWriter, we can generate parquet/json files and upload the files to milvus s3 storage:
```
    with RemoteBulkWriter(
            schema=schema,
            remote_path="bulk_data",
            connect_param=RemoteBulkWriter.S3ConnectParam(
                endpoint=MINIO_ADDRESS,
                access_key=MINIO_ACCESS_KEY,
                secret_key=MINIO_SECRET_KEY,
                bucket_name="a-bucket",
            ),
            segment_size=512 * 1024 * 1024,
            file_type=BulkFileType.PARQUET,
    ) as remote_writer:
        for i in range(10000):
            remote_writer.append_row({"id": i, "vector": gen_float_vector()})

        remote_writer.commit()
        batch_files = remote_writer.batch_files
        print(batch_files)
```
"connect_param" defines the S3 address/port/bucket of the milvus
"remote_path" defines the directory where to upload the parquet/json files
"segment_size" defines how large per file, if the data size exceeds this value, there will be multiple files generated

After the RemoteBulkWriter finishes work, the print(batch_files) will print out the paths of the files on the S3.

Since the "remote_path" is "bulk_data", the path will look like:
"/bulk_data/31ad591e-97c9-42c6-9c68-94daa43eea8a/1.parquet"

Then you can use this path to call the import api:
```
curl --request POST "http://xxxxx:19530/v2/vectordb/jobs/import/create"
--header "Content-Type: application/json"
--data-raw '{
"files": [
  [
    "/bulk_data/31ad591e-97c9-42c6-9c68-94daa43eea8a/1.parquet"
  ]
],
"collectionName": "quick_setup"
}
```

**Reply by liorf95** [reply]
Using RemoteBulkWriter already works perfectly for me.
The issue is when I try to use LocalBulkWriter to create the parquet files locally, scp it to the remote minio bulk_data mounted path, and then call the import API.

**Comment by liorf95** [comment]
@xiaofan-luan 

Can you please provide details on where the import API code of the Milvus server is exactly?
Is it on https://github.com/milvus-io/milvus/tree/master/internal/core/src or some other path?

**Reply by yhmo** [reply]
The import rpc entry of proxy: https://github.com/milvus-io/milvus/blob/3685edb264c2f4de691d35a719555c05968d40d8/internal/proxy/impl.go#L4757

The import interface of datacoord: https://github.com/milvus-io/milvus/blob/3685edb264c2f4de691d35a719555c05968d40d8/internal/datacoord/services.go#L1666
And the import_job,go/import_scheduler.go/import_task.go/import_meta.go under the path /milvus/blob/master/internal/datacoord to manage the bulkinsert jobs. Datacoord splits the job into tasks and assigns them to data nodes.

The import interface of datanode: https://github.com/milvus-io/milvus/blob/3685edb264c2f4de691d35a719555c05968d40d8/internal/datanode/services.go#L470
The source code to manage bulkinsert tasks in datanode:
https://github.com/milvus-io/milvus/tree/master/internal/datanode/importv2

The major code to parse the json/parquet/numpy file.
https://github.com/milvus-io/milvus/tree/master/internal/util/importutilv2

**Reply by liorf95** [reply]
That info is great for me! Thanks a lot!

**Comment by xiaofan-luan** [comment]
bulkinsert is a complicated feature, it's under  https://github.com/milvus-io/milvus/tree/master/internal

to start from, you need to read proxy ImportV2 API

The process is like: 
proxy send request to datacoord
datacoord create an import task and schedule it to datanode
datanode execute with several steps

**Reply by liorf95** [reply]
Thanks! This is very helpful for me!
Is any part of the import task being parallel/multithreaded? 
For example, if we have 3 parquet files to import and 3 data nodes on the cluster - will those files be written to the Object Storage in a parallel way all 3 together at once each file on a different datanode, and not in a serial way one after the other?
